packages/uploadthing/src/sdk/utils.ts

```diff
@@ -76,88 +76,135 @@ export const uploadFilesInternal = async (
+      const { presignedUrls, key, fileUrl, uploadId, chunkSize } = json.data[i];
 
-      if (!presignedUrl || !fields) {
+      if (!presignedUrls || !Array.isArray(presignedUrls)) {
         throw new UploadThingError({
           code: "URL_GENERATION_FAILED",
           message: "Failed to generate presigned URL",
           cause: JSON.stringify(json.data[i]),
         });
       }
 
-      const formData = new FormData();
-      formData.append("Content-Type", file.type);
-      Object.entries(fields).forEach(([key, value]) => {
-        formData.append(key, value);
-      });
-
-      formData.append(
-        "file",
-        // Handles case when there is no file name
-        file.name
-          ? (file as File)
-          : Object.assign(file as File, { name: "unnamed-blob" }),
-      );
-
-      // Do S3 upload
-      const s3res = await opts.fetch(presignedUrl, {
-        method: "POST",
-        body: formData,
-        headers: new Headers({
-          Accept: "application/xml",
+      const maxRetries = 10;
+      const etags = await Promise.all(
+        presignedUrls.map(async (url, partNumber) => {
+          let retryCount = 0;
+
+          const offset = chunkSize * partNumber;
+          const end = Math.min(offset + chunkSize, file.size);
+          const chunk = file.slice(offset, end);
+
+          async function makeRequest() {
+            const s3Res = await opts.fetch(url, {
+              method: "PUT",
+              body: chunk as Blob, // omit weird union of different blobs
+              headers: {
+                "Content-Type": file.type,
+                "Content-Disposition": [
+                  data.contentDisposition,
+                  `filename="${file.name}"`,
+                  `filename*=UTF-8''${file.name}`,
+                ].join("; "),
+              },
+            });
+
+            if (s3Res.ok) {
+              const etag = s3Res.headers.get("Etag");
+              if (!etag) {
+                throw new UploadThingError({
+                  code: "UPLOAD_FAILED",
+                  message: "Missing Etag header from uploaded part",
+                });
+              }
+
+              console.log(`Uploaded part ${partNumber} with etag ${etag}`);
+              return etag.replace(/"/g, "");
+            }
+
+            if (retryCount < maxRetries) {
+              retryCount++;
+              // Retry after exponential backoff
+              const delay = 2 ** retryCount * 1000;
+              await new Promise((r) => setTimeout(r, delay));
+              console.log(
+                `Retrying part ${partNumber} after ${delay}ms, attempt ${
+                  retryCount + 1
+                }/${maxRetries}`,
+              );
+              return makeRequest();
+            }
+
+            // Max retries exceeded, tell UT server that upload failed
+            await opts.fetch(generateUploadThingURL("/api/failureCallback"), {
```

Comment:
should we include a prop that this is multipart here, and then on the infra side we can handle that and call `AbortIncompleteMultipartUpload ` ?